### ISSUE TITLE
`UI::HtmlDialog#show` for an open window: sets focus on Mac, not on Windows

### ISSUE THREAD

**Issue by DanRathbun** [SketchUp/api-issue-tracker#647]
####  SketchUp Ruby API Issue

**`UI::HtmlDialog#show` for an open window: sets focus on Mac, not on Windows.**

This is platform parity issue.

SketchUp Version: `2021.1.279`

```ruby
wd = UI::HtmlDialog.new(
{
  :dialog_title => "ChromeEmbeddedFramework",
  :style => UI::HtmlDialog::STYLE_DIALOG
})
wd.set_url('chrome://version/')
wd.show
```

From the console (which gives it focus):
```ruby
wd.bring_to_front # works
wd.show # does NOT work on Windows
```

From a menu item (which gives the main application window focus):
```ruby
UI.menu("Plugins").add_item("Test bring_to_front") { wd.bring_to_front }
UI.menu("Plugins").add_item("Test show") { wd.show }
```
Again, `#bring_to_front` works and `#show` does **not** on Windows.

----

I'm not sure what should happen when [`#show_modal`](https://ruby.sketchup.com/UI/HtmlDialog.html#show_modal-instance_method) is called upon an **open** non-modal dialog. (At the very least transparently call [`#bring_to_front`](https://ruby.sketchup.com/UI/HtmlDialog.html#bring_to_front-instance_method), but I'm not sure of the use case for changing an open window's modality.)

----

> _Previously discussed in [Issue 494](https://github.com/SketchUp/api-issue-tracker/issues/494#issuecomment-860222948), where I said:_
>
> > > @DanRathbun : Does calling `#show` on a visible dialog also give it focus ?
> > 
> > 
> > @Fredosixx : It does on Mac, but not on Windows. Yet another inconsistency....
> 
> **Is this still true for the latest 2021.1 release ?**
> 
> Should we expect that `#show` would give a visible dialog the focus.
> Ie, when visible, **should** calling `#show` be a wrapper for `#bring_to_front`.

... _and Thomas replied thus:_
> @thomthom : `#show` and `#bring_to_front` should now on both platform consistently set focus to the dialog.

~

**Comment by thomthom** [comment]
```
wd = UI::HtmlDialog.new(
{
  :dialog_title => "ChromeEmbeddedFramework",
  :style => UI::HtmlDialog::STYLE_DIALOG
})
wd.set_url('chrome://version/')
wd.show
```

This seems to give focus to the dialog on my machine with SU2021.1.

The dialog opens and I can immediately use the arrow keys to scroll the window.

**Comment by DanRathbun** [comment]
> This seems to give focus to the dialog on my machine with SU2021.1.
> The dialog opens and I can immediately use the arrow keys to scroll the window.

I think you misunderstand the issue, ... the title clearly says "for an open window", .. ie, a dialog that is open but does **not** have the focus.

In the first post I showed and said that when **using the console** it (the console) would gain the focus and the dialog would loose it.

**Comment by thomthom** [comment]
Oh... calling `wd.show` a second time while the window is already open. That we didn't define a behaviour for - in terms of focus.

**Comment by DanRathbun** [comment]
Yes, but Fredo reports that on Mac the window is refocused, on Windows it's a no-op.

**Comment by Fredosixx** [comment]
I reported the problem for the `bring_to_front` method. This has been fixed in SU2021.1 for Windows.

`show` is a one-time method to show the html dialog for the first time. I don't use it afterward, because it does not do anything when the dialog is already on screen.

**Comment by DanRathbun** [comment]
You indicated that on Mac it would set the focus to an already open window. **Is this not true ?**

**Comment by Fredosixx** [comment]
the `show` method should not be used after the initial display. It is the method 'bring_to_front' that puts the keyboard focus, correctly on Mac and NOT on Windows before SU2021.1. 

It just happens that `show` on Windows does not do put the focus either.

**Comment by DanRathbun** [comment]
**The point here is about parity between the platforms.**

It is **probable** that some coders might call `#show` on an open dialog without coding a conditional that checks it's visibility and instead calls `#bring_to_front`.

If this paradigm on Mac works, and it is not too difficult, then on Windows the same should also be made to work (ie, `#show` would transparently call `#bring_to_front` or otherwise give the dialog UI focus.)

**Comment by DanRathbun** [comment]
What I'm advocating is, that since [`UI::HtmlDialog#show`](https://ruby.sketchup.com/UI/HtmlDialog.html#show-instance_method) on MAC already works like [`UI::Toolbar#restore`](https://ruby.sketchup.com/UI/Toolbar.html#restore-instance_method), then this should be a **parity** issue and it should be made to work the same on WIN.

@thomthom I don't think you need more info. It's **parity**, although perhaps not a bug but an **enhancement**.

We might **also** consider the oft situation where dialogs get "lost offscreen" and the main application seems to go unresponsive causing the user to think SketchUp has "locked up." We could entertain the idea that calling `#show` on an open dialog should be a demand that the dialog window be moved within the viewable area of the display it should be displayed on.
